### PR TITLE
[SPARK-33193][SQL][TEST] Hive ThriftServer JDBC Database MetaData API Behavior Auditing

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

Add a test case to audit all JDBC metadata behaviors to check and prevent potential APIs silent changing from both the upstream hive-jdbc module or the Spark thrift server side.

Forked from my kyuubi project here https://github.com/yaooqinn/kyuubi/blob/master/externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/operation/SparkOperationSuite.scala

### Why are the changes needed?

Make the SparkThriftServer safer to evolve.

### Does this PR introduce _any_ user-facing change?

dev only

### How was this patch tested?

new tests

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan @wangyum @juliuszsompolski @maropu @dongjoon-hyun @HyukjinKwon thanks ~

**Comment by juliuszsompolski** [comment]
cc @bogdanghit

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala [hunk 1]
```diff
@@ -396,4 +400,187 @@ class SparkMetadataOperationSuite extends HiveThriftJdbcTest {
       }
     }
   }
+
+  test("Hive ThriftServer JDBC Database MetaData API Auditing") {
```
Comment (maropu) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala:404]: How about splitting this test into the two parts?
```
test("Hive ThriftServer JDBC Database MetaData API Auditing - supported") {
test("Hive ThriftServer JDBC Database MetaData API Auditing - not supported") {
```

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala [hunk 2]
```diff
@@ -396,4 +400,187 @@ class SparkMetadataOperationSuite extends HiveThriftJdbcTest {
...
+        () => metaData.getURL,
+        () => metaData.getUserName,
+        () => metaData.isReadOnly,
+        () => metaData.nullsAreSortedHigh,
+        () => metaData.nullsAreSortedLow,
+        () => metaData.nullsAreSortedAtStart(),
+        () => metaData.nullsAreSortedAtEnd(),
+        () => metaData.usesLocalFiles(),
+        () => metaData.usesLocalFilePerTable(),
+        () => metaData.supportsMixedCaseIdentifiers(),
```
Comment (maropu) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala:418]: nit: we don't need `()` here: https://github.com/databricks/scala-style-guide#parentheses

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala [hunk 3]
```diff
@@ -396,4 +400,187 @@ class SparkMetadataOperationSuite extends HiveThriftJdbcTest {
...
+        () => metaData.supportsANSI92EntryLevelSQL,
+        () => metaData.supportsANSI92IntermediateSQL,
+        () => metaData.supportsANSI92FullSQL,
+        () => metaData.supportsIntegrityEnhancementFacility,
+        () => metaData.isCatalogAtStart,
+        () => metaData.supportsSubqueriesInComparisons,
+        () => metaData.supportsSubqueriesInExists,
+        () => metaData.supportsSubqueriesInIns,
+        () => metaData.supportsSubqueriesInQuantifieds,
+        // Spark support this, see https://issues.apache.org/jira/browse/SPARK-18455
```
Comment (maropu) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala:449]: This comment looks a bit confusing. btw, could we fix this in followup activities? If we can, could you file it in jira?

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala (hunk 1 above)
Comment (yaooqinn) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala:404]: SGTM

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala (hunk 3 above)
Comment (yaooqinn) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala:449]: we can't, it belongs to the upstream hive module where defines the hive behavior.  Here we just highlight the difference and make it perspective in future changes

Comment (maropu) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala:449]: Ah, I see...

Comment (maropu) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala:449]: If so, could you write it in the comment like that?

**Comment by maropu** [comment]
Looks okay otherwise.

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala (hunk 3 above)
Comment (yaooqinn) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala:449]: yea~

**Comment by yaooqinn** [comment]
retest this please

**Comment by yaooqinn** [comment]
retest this please

**Review by cloud-fan — APPROVED** [review]

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala [hunk 4]
```diff
@@ -396,4 +400,203 @@ class SparkMetadataOperationSuite extends HiveThriftJdbcTest {
...
+      // This is actually supported, but hive jdbc package return false
+      assert(!metaData.supportsUnion)
+      assert(metaData.supportsUnionAll)
+      assert(metaData.getMaxColumnNameLength === 128)
+      assert(metaData.getDefaultTransactionIsolation === java.sql.Connection.TRANSACTION_NONE)
+      assert(!metaData.supportsTransactions)
+      assert(!metaData.getProcedureColumns("", "%", "%", "%").next())
+      assert(!metaData.getImportedKeys("", "default", "").next())
+
+      // We should disable these two APIs
```
Comment (dongjoon-hyun) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala:585]: Could you file JIRAs for these two?

Comment (dongjoon-hyun) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala:585]: BTW, is it feasible?

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala [hunk 5]
```diff
@@ -396,4 +400,203 @@ class SparkMetadataOperationSuite extends HiveThriftJdbcTest {
...
+      assert(!metaData.getPrimaryKeys("", "default", "").next())
+      assert(!metaData.getCrossReference("", "default", "src", "", "default", "src2").next())
+
+      assert(!metaData.getIndexInfo("", "default", "src", true, true).next())
+      assert(metaData.supportsResultSetType(new Random().nextInt()))
```
Comment (dongjoon-hyun) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala:590]: Why do we use `Random` here? For API auditing, it seems that we can use one of the valid one.
If you want to test all ranges of input domain, this `Random` usage means a flaky test.

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala [hunk 6]
```diff
@@ -396,4 +400,203 @@ class SparkMetadataOperationSuite extends HiveThriftJdbcTest {
...
+
+  test("Hive ThriftServer JDBC Database MetaData API Auditing - Method supported") {
+    // These APIs belong to the upstream Apache Hive's hive-jdbc artifact where defines the hive
+    // behavior. Users can also use it to interact with Spark ThriftServer directly. Some behaviors
+    // are not fully consistent with Spark e.g. we can work with multiple catalogs.
+    // There is nothing we can do but accept the current condition and highlight the difference
+    // and make it perspective in future changes both from upstream and inside Spark.
+    withJdbcStatement() { statement =>
+      val metaData = statement.getConnection.getMetaData
+      import org.apache.spark.SPARK_VERSION
```
Comment (dongjoon-hyun) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala:531]: If there is no reason, shall we move this `import` at the header of this file?

**Comment by dongjoon-hyun** [comment]
Thank you, @yaooqinn . This is very useful. I left a few minor comments.

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala (hunk 4 above)
Comment (yaooqinn) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SparkMetadataOperationSuite.scala:585]: yes, it is. e.g.  https://github.com/yaooqinn/kyuubi/blob/master/kyuubi-common/src/main/scala/org/apache/kyuubi/service/FrontendService.scala#L369-L383
I will file a ticket and mark TODO here

**Comment by dongjoon-hyun** [comment]
Thank you, @yaooqinn and all.
Merged to master for Apache Spark 3.1+ test coverage.

**Comment by HyukjinKwon** [comment]
+1, nice.

**Pull request closed without merging**